### ISSUE TITLE
Consider doing all development in branches and PRs

### ISSUE THREAD

**Issue by osma** [adbar/simplemma#37]
Hi @adbar, just a small suggestion on development practices: it seems to me that you are adding commits directly to the `main` branch when doing work like [this](https://github.com/adbar/simplemma/pull/23#issuecomment-1380882178). It is of course up to you how you want to work on the codebase (hey it's your project!). I used to work that way too, but nowadays prefer to use branches and then open PRs, even for many personal projects. There's a little more housekeeping to take care of with all the branches and PRs, but the benefits are clear:

- it's easier to see what is currently being worked on (for example in draft PRs) and thus easier for contributors to chime in
- later it's also easier to know what features were worked on for example in a particular release, the code changes can all be found in a single place, possibly with related discussion in the comments
- if you use CI and/or code quality analysis tools (there are many great ones that are free for Open Source projects, for example SonarCloud and Code Climate), you get near-instant feedback about potential issues in the code and ways to refactor or otherwise improve it
- the workflow is the same as with PRs coming from external contributors, making it simpler to understand what's going on
- perhaps the most important: it helps to ensure that the main branch is always in working condition - don't merge any work to `main` until you're sure that the quality is good!

Thanks for considering, and feel free to close this regardless of what you decide to do.

**Comment by adbar** [comment]
Hi @osma, thanks for making this point. Each commit follows a series of formal checks (black, mypy) and I wouldn't push anything that degrades performance on the UD benchmark.
But I agree with you, the repository has gotten more attention recently and using draft PRs allow for a clearer structure for everyone involved.

**Comment by osma** [comment]
Thanks for your thoughts. Feel free to close this. I won't do it myself, in case you want to keep this open as some sort of reminder.

Also consider using some of the free quality analysis tools that are available for scanning PRs. For Annif, we currently use Github CodeQL (formerly LGTM.com), SonarCloud, Code Climate, Scrutinizer and Codecov. All are easy to set up and they tend to report slightly different issues, so complement each other. Out of these, you seem to use only Codecov.

**Comment by adbar** [comment]
I use flake8, black, mypy and Github CodeQL. Given that this is (for now) a rather small code base I think we're covered, but thanks for the suggestion!
